exonum-java-proofs/src/main/java/com/exonum/binding/storage/proofs/map/flat/UncheckedFlatMapProof.java

```diff
@@ -0,0 +1,201 @@
+package com.exonum.binding.storage.proofs.map.flat;
+
+import static com.exonum.binding.hash.Funnels.hashCodeFunnel;
+import static com.exonum.binding.storage.proofs.DbKeyFunnel.dbKeyFunnel;
+import static java.util.stream.Collectors.toList;
+
+import com.exonum.binding.hash.HashCode;
+import com.exonum.binding.hash.HashFunction;
+import com.exonum.binding.hash.Hashing;
+import com.exonum.binding.storage.proofs.map.DbKey;
+import com.exonum.binding.storage.proofs.map.DbKey.Type;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * An unchecked flat map proof, which does not include any intermediate nodes.
+ */
+public class UncheckedFlatMapProof implements UncheckedMapProof {
+
+  private static final HashFunction HASH_FUNCTION = Hashing.defaultHashFunction();
+
+  private final List<MapProofEntry> proofList = new ArrayList<>();
+
+  private final List<MapProofEntry> proof;
+
+  private final List<MapEntry> entries;
+
+  private final List<byte[]> missingKeys;
+
+  UncheckedFlatMapProof(
+      List<MapProofEntry> proof,
+      List<MapEntry> entries,
+      List<byte[]> missingKeys) {
+    this.proof = proof;
+    this.entries = entries;
+    this.missingKeys = missingKeys;
+  }
+
+  @SuppressWarnings("unused") // Native API
+  static UncheckedFlatMapProof fromUnsorted(
+      MapProofEntry[] proofList,
+      MapEntry[] entries,
+      byte[][] missingKeys) {
+    List<MapProofEntry> proof = Arrays.asList(proofList);
+    List<MapEntry> entriesList = Arrays.asList(entries);
+    List<byte[]> missingKeysList = Arrays.asList(missingKeys);
+    return new UncheckedFlatMapProof(proof, entriesList, missingKeysList);
+  }
+
+  @Override
+  public CheckedMapProof check() {
+    if (prefixesIncluded()) {
+      return CheckedFlatMapProof.invalid(ProofStatus.INVALID_STRUCTURE);
+    }
+    ProofStatus orderCheckResult = orderCheck();
+    if (orderCheckResult != ProofStatus.CORRECT) {
+      return CheckedFlatMapProof.invalid(orderCheckResult);
+    }
+    mergeLeavesWithBranches();
+    if (proofList.isEmpty()) {
+      return CheckedFlatMapProof.correct(
+          getEmptyProofListHash(), Collections.emptyList(), missingKeys);
+    } else if (proofList.size() == 1) {
+      // One element proof is correct only if the node is a leaf
+      if (proof.size() == 1 && proof.get(0).getDbKey().getNodeType() != Type.LEAF) {
```

Comment:
This code is rather hard to understand. It also seems more complicated than it is:
![image](https://user-images.githubusercontent.com/7531824/45025387-056b8f80-b044-11e8-9ddc-699cb9b1aea5.png)

I’d go with something like that:
```java

  private CheckedMapProof checkSingletonProof() {
    assert proof.size() + entries.size() == 1;

    if (proof.size() == 1) {
      // There are no entries, therefore, the proof node must correspond to a leaf.
      MapProofEntry entry = proof.get(0);
      DbKey.Type nodeType = entry.getDbKey().getNodeType();
      if (nodeType == Type.BRANCH) {
        return CheckedFlatMapProof.invalid(ProofStatus.NON_TERMINAL_NODE);
      } else {
        // TODO: calculate hash properly.
      }
    } else {
      // TODO: calculate hash properly.
    }
  }

```

It also must be decided where we calculate the branch (root) hash: here or determine in `merge…`. In the first case we have a merge that does not work correctly for singleton maps, doing meaningless operations. In the second case we have a rather complicated merge method.

Probably, it would be better to refactor in the following way:
```
def check:
  checkNoPrefixes
  checkOrder
  if isEmptyTree():
    return checkEmptyTree()
  elif isSingletonTree():
    return checkSingletonTree()
  else:
    return checkTree() // Performs merge internally
```

The first case (calculating it here after meaningless merge) is also acceptable.